ui/ozone/platform/x11/x11_display_manager_ozone.cc

```diff
@@ -23,34 +24,36 @@ namespace {
 
 constexpr int default_refresh = 60;
 
-X11DisplayManagerOzone::Snapshot CreateSnapshot(int64_t display_id,
-                                                gfx::Rect bounds,
-                                                gfx::ColorSpace color_space) {
-  std::unique_ptr<display::DisplaySnapshot> snapshot =
-      std::make_unique<display::DisplaySnapshot>(
-          display_id, gfx::Point(bounds.x(), bounds.y()),
-          gfx::Size(bounds.width(), bounds.height()),
-          display::DisplayConnectionType::DISPLAY_CONNECTION_TYPE_NONE, false,
-          false, false, color_space, "", base::FilePath(),
-          display::DisplaySnapshot::DisplayModeList(), std::vector<uint8_t>(),
-          nullptr, nullptr, 0, 0, gfx::Size());
-  std::unique_ptr<display::DisplayMode> current_mode =
+std::unique_ptr<display::DisplaySnapshot> CreateSnapshot(
+    int64_t display_id,
+    gfx::Rect bounds,
+    gfx::ColorSpace color_space) {
+  display::DisplaySnapshot::DisplayModeList modes;
+  std::unique_ptr<display::DisplayMode> display_mode =
       std::make_unique<display::DisplayMode>(
           gfx::Size(bounds.width(), bounds.height()), false, default_refresh);
-  snapshot->set_current_mode(current_mode.get());
-  return std::make_pair(std::move(snapshot), std::move(current_mode));
+  modes.push_back(std::move(display_mode));
+  const display::DisplayMode* mode = modes.back().get();
```

Comment:
|display_mode| is owned by |modes|. Where |modes| are only defined within the scope of this method. This shall result in undefined behavior later. Why can't you just store the current |modes| and the snapshots as WaylandOutput does?

Reply:
When it creates display::DisplaySnapshot,  it deliver |modes| as a parameter and DisplaySnapshot keeps it. I refereed to others, how to create display::DisplaySnapshot, https://cs.chromium.org/chromium/src/ui/display/mojo/display_snapshot_struct_traits.cc?sq=package:chromium&g=0&l=131 or https://cs.chromium.org/chromium/src/ui/ozone/platform/drm/common/drm_util.cc?sq=package:chromium&q=drm_uti&g=0&l=472.

And whenever DisplaySnapshot::add_mode() is called, it clones every display mode and keeps it.

FYI, I think we should have this mechanism for headless as well and I was supposed to upstream it after confirm this change but it's already reverted today with the same reason. https://chromium-review.googlesource.com/c/chromium/src/+/1094851/5/ui/ozone/platform/headless/headless_native_display_delegate.cc#35

Reply:
ah, I haven't noticed it has also taken modes.